### PR TITLE
Discovery2319

### PR DESCRIPTION
**Pull request by johnataylor**
Can you have a quick look. There are three commits here, for the three issues. And this works and doesn't appear to regress. We can move the extra http into a separate resource.

@yishaigalatzer 
@joelverhagen 
@emgarten

### CODE REVIEW

File: submodules/Common [hunk 1]
```diff
@@ -1 +1 @@
-Subproject commit e6fac8061686c18531e2621ccef97dd5e0687a65
```
Comment (joelverhagen) [inline submodules/Common]: Are these submodule changes intentional?

Comment (emgarten) [inline submodules/Common]: I just updated those today, that shouldn't be changed. Maybe this wasn't rebased.

File: src/NuGet.Core/NuGet.Protocol.Core.v3/HttpSource/HttpSource.cs [hunk 2]
```diff
@@ -171,6 +172,19 @@ public HttpSource(PackageSource source, Func<Task<HttpHandlerResource>> messageH
...
 
+        public Task<HttpResponseMessage> GetAsync(Uri uri, MediaTypeWithQualityHeaderValue[] accept, ILogger log, CancellationToken token)
+        {
+            var request = new HttpRequestMessage(HttpMethod.Get, uri);
+            foreach (var a in accept)
+            {
+                request.Headers.Accept.Add(a);
+            }
+
+            Func<HttpRequestMessage> requestFactory = () => request;
```
Comment (davidfowl) [inline src/NuGet.Core/NuGet.Protocol.Core.v3/HttpSource/HttpSource.cs:183]: Shouldn't this be:

``` C#
Func<HttpRequestMessage> requestFactory = () =>
{
    var request = new HttpRequestMessage(HttpMethod.Get, uri);
    foreach (var a in accept)
    {
        request.Headers.Accept.Add(a);
    }
    return request;
};
```

File: src/NuGet.Core/NuGet.Protocol.Core.v3/HttpSource/HttpSource.cs [hunk 3]
```diff
@@ -19,6 +19,7 @@
 using NuGet.Logging;
 using NuGet.Protocol.Core.Types;
 using NuGet.Protocol.Core.v3;
+using System.Net.Http.Headers;
```
Comment (emgarten) [inline src/NuGet.Core/NuGet.Protocol.Core.v3/HttpSource/HttpSource.cs:22]: nit: sort

File: src/NuGet.Core/NuGet.Protocol.Core.v3/LegacyFeed/ODataServiceDocumentResourceV2Provider.cs [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+
+            var cacheContext = HttpSourceCacheContext.CreateCacheContext(new SourceCacheContext(), 0);
+
+            HttpSourceResult response = await client.GetAsync(
+                url,
+                "odata_service_document",
+                cacheContext,
+                log,
+                ignoreNotFounds: false,
+                ensureValidContents: null,
```
Comment (emgarten) [inline src/NuGet.Core/NuGet.Protocol.Core.v3/LegacyFeed/ODataServiceDocumentResourceV2Provider.cs:100]: it wouldn't hurt to validate that this is xml

**Comment by emgarten** [comment]
I suggest verifying how offline and failing v2 sources are handled with this change, it may fail sooner than it did before and with a different message. Overall this will help a lot to determine if the source really is v2 before sending requests.

I would expect this to affect the network call count tests, it may pass on the second run due to disk caching. This was tricky to fix for index.json caching but you can follow the same pattern.

Fix issues, then :shipit:

File: src/NuGet.Core/NuGet.Protocol.Core.v3/LegacyFeed/ODataServiceDocumentResourceV2Provider.cs (hunk 4 above)
Comment (joelverhagen) [inline src/NuGet.Core/NuGet.Protocol.Core.v3/LegacyFeed/ODataServiceDocumentResourceV2Provider.cs:100]: Use `HttpStreamValidation.ValidateXml`.

**Comment by johnataylor** [comment]
We've split this pull request and already merged the Accept header change and the TargetFramework change. The remaining changes are all around the introduction of a resource to handle the base address correctly. Handling the base address correctly fixes the issues we have with redirects.

**Comment by emgarten** [comment]
Should this go into `dev` or what is the plan for this?

**Comment by yishaigalatzer** [comment]
This should be in dev

**Pull request closed without merging**